browser/net/brave_site_hacks_network_delegate_helper.cc

```diff
@@ -125,25 +97,31 @@ void ApplyPotentialQueryStringFilter(std::shared_ptr<BraveRequestInfo> ctx) {
     return;
   }
 
-  // Regular expressions are dirty and error-prone, but unfortunately
-  // there is no right way to parse a query string, other than one
-  // generated by a URL-encoded HTML form submission. See
-  // https://github.com/brave/brave-core/pull/3239#issuecomment-524073918
-  // for more information on why this approach was selected.
-
-  std::string new_query = ctx->request_url.query();
-  // Note: the ordering of these replacements is important.
-  const int replacement_count =
-      re2::RE2::GlobalReplace(&new_query, tracker_appended_matcher.Get(), "") +
-      re2::RE2::GlobalReplace(&new_query, tracker_first_matcher.Get(), "") +
-      re2::RE2::GlobalReplace(&new_query, tracker_only_matcher.Get(), "");
-
-  if (replacement_count > 0) {
+  const std::string query = ctx->request_url.query();
```

Comment:
can we please recheck that `net::QueryIterator` is still not suitable for us? Manual parsing doesn't look super appealing

Reply:
Good suggestion, @iefremov.

I looked at `net::QueryIterator` carefully, and tested it, and confirmed what Francois reported [on Aug 22, 2019](https://github.com/brave/brave-core/pull/3239#issuecomment-524073918). That is: `net::QueryIterator` parser, which calls [url::ExtractQueryKeyValue()](https://cs.chromium.org/chromium/src/url/third_party/mozilla/url_parse.cc?l=654&rcl=1ddb53ff33a740e30de6dbe4cfe13b1802e69e54) makes assumptions about keys and values that will result in modifying the URL unnecessarily.

The logic I am using (though not the implementation) is the same as Francois's and fairly simple.
1. Split the query string by ampersands (`&`). Each fragment (even empty ones) is a kvString.
2. Split each kvString at the first equal sign (`=`). Whatever comes before (even an empty string) is the key; everything afterwards is the value.
3. Remove any kvString whose key corresponds to a tracking key. (For this patch, check for conditional stripping as well.)
4. Join the remaining kvStrings back together with ampersands, to create the sanitized query string.

My new implementation, passes all existing unit and browser tests with no modifications. (In addition I added new tests to verify the behavior of our `mkt_tok` rule.)

Reply:
gotcha, can we at least extract this code to a separate function?

Reply:
done